File: integration_tests/dart_client_tests_real/test/mutations_test.dart
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'package:graphlink_client_integration_tests/generated/enums/user_status.dart';
import 'package:graphlink_client_integration_tests/generated/enums/priority.dart';
import 'package:graphlink_client_integration_tests/generated/inputs/create_user_input.dart';
import 'package:graphlink_client_integration_tests/generated/inputs/update_user_input.dart';
import 'package:graphlink_client_integration_tests/generated/inputs/address_input.dart';
import 'real_server_adapter.dart';

// The server echoes input fields back — priority is null when not in input.
final _minimalInput = CreateUserInput(
  name: 'Alice Smith',
  email: 'alice@test.com',
  status: UserStatus.ACTIVE,
  address:
      AddressInput(street: '123 Main St', city: 'Springfield', country: 'US'),
);

void main() {
  late GraphLinkClient client;
  setUp(() => client = newClient());

  // ── createUser ────────────────────────────────────────────────────────────

  group('createUser — scalar fields echoed from input', () {
    test('name is returned from server', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.name, equals('Alice Smith'));
    });

    test('email is returned from server', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.email, equals('alice@test.com'));
    });

    test('status is returned as enum', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.status, equals(UserStatus.ACTIVE));
    });

    test('priority is null when not provided in input', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.priority, isNull);
    });

    test('priority is deserialized as enum when provided', () async {
      final res = await client.mutations.createUser(
        input: CreateUserInput(
          name: 'Alice Smith',
          email: 'alice@test.com',
          status: UserStatus.ACTIVE,
          priority: Priority.HIGH,
          address: AddressInput(
              street: '123 Main St', city: 'Springfield', country: 'US'),
        ),
        limit: 10,
      );
      expect(res.createUser.priority, equals(Priority.HIGH));
    });
  });

  group('createUser — nested address from input', () {
    test('address.street is returned', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.address.street, equals('123 Main St'));
    });

    test('address.city is returned', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.address.city, equals('Springfield'));
    });

    test('address.zip is null when not provided', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.address.zip, isNull);
    });

    test('nullable billingAddress is null when not provided', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.billingAddress, isNull);
    });
  });

  group('createUser — list fields from input', () {
    test('tags list is echoed back from input', () async {
      final res = await client.mutations.createUser(
        input: CreateUserInput(
          name: 'Alice Smith',
          email: 'alice@test.com',
          status: UserStatus.ACTIVE,
          address: AddressInput(
              street: '123 Main St', city: 'Springfield', country: 'US'),
          tags: ['admin', 'beta'],
        ),
        limit: 10,
      );
      expect(res.createUser.tags, equals(['admin', 'beta']));
    });

    test('tags defaults to empty list when not provided', () async {
      final res =
          await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.tags, isEmpty);
    });
  });

  // ── deleteUser ────────────────────────────────────────────────────────────

  group('deleteUser', () {
    test('returns true', () async {
      final res = await client.mutations.deleteUser(id: 'user-1');
      expect(res.deleteUser, isTrue);
    });
  });

  // ── updateUser ────────────────────────────────────────────────────────────

  group('updateUser', () {
    test('updated name is reflected in response', () async {
      final res = await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(name: 'Updated Alice'),
        limit: 10,
      );
      expect(res.updateUser.id, equals('user-1'));
      expect(res.updateUser.name, equals('Updated Alice'));
    });

    test('unchanged fields are preserved from base user', () async {
      final res = await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(name: 'New Name'),
        limit: 10,
      );
      expect(res.updateUser.email, equals('alice@test.com'));
      expect(res.updateUser.status, equals(UserStatus.ACTIVE));
    });

    test('status update is applied', () async {
      final res = await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(status: UserStatus.SUSPENDED),
        limit: 10,
      );
      expect(res.updateUser.status, equals(UserStatus.SUSPENDED));
    });

    test('nested address is preserved', () async {
      final res = await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(name: 'New Name'),
        limit: 10,
      );
      expect(res.updateUser.address.street, equals('123 Main St'));
    });
  });
}

File: integration_tests/typescript_client_tests_real/test/mutations.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import { UserStatus } from '../lib/generated/enums/user-status.js';
import { Priority } from '../lib/generated/enums/priority.js';
import type { CreateUserInput } from '../lib/generated/inputs/create-user-input.js';
import type { UpdateUserInput } from '../lib/generated/inputs/update-user-input.js';
import { newClient } from './real-server-adapter.ts';

const minimalInput: CreateUserInput = {
  name: 'Alice Smith',
  email: 'alice@test.com',
  status: UserStatus.Active,
  address: { street: '123 Main St', city: 'Springfield', country: 'US' },
};

let client: GraphLinkClient;
beforeEach(() => { client = newClient(); });

// ── createUser ────────────────────────────────────────────────────────────────

describe('createUser — scalar fields', () => {
  it('id is user-new', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.id).toBe('user-new');
  });

  it('name is echoed', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.name).toBe('Alice Smith');
  });

  it('email is echoed', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.email).toBe('alice@test.com');
  });

  it('status is deserialized as enum', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.status).toBe(UserStatus.Active);
  });

  it('priority is null when not in input', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.priority).toBeNull();
  });
});

describe('createUser — with optional fields', () => {
  it('priority is echoed when provided', async () => {
    const res = await client.mutations.createUser({ input: { ...minimalInput, priority: Priority.High } });
    expect(res.createUser.priority).toBe(Priority.High);
  });

  it('status INACTIVE is echoed', async () => {
    const res = await client.mutations.createUser({ input: { ...minimalInput, status: UserStatus.Inactive } });
    expect(res.createUser.status).toBe(UserStatus.Inactive);
  });
});

describe('createUser — nested address', () => {
  it('address fields are echoed', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.address.street).toBe('123 Main St');
    expect(res.createUser.address.city).toBe('Springfield');
    expect(res.createUser.address.country).toBe('US');
  });

  it('nullable billingAddress is null when not provided', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.billingAddress).toBeNull();
  });
});

describe('createUser — list fields', () => {
  it('tags is empty when not provided', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.tags).toHaveLength(0);
  });
});

// ── deleteUser ────────────────────────────────────────────────────────────────

describe('deleteUser — returns bool', () => {
  it('returns true', async () => {
    const res = await client.mutations.deleteUser({ id: 'user-1' });
    expect(res.deleteUser).toBe(true);
  });

  it('returns true regardless of id', async () => {
    const res = await client.mutations.deleteUser({ id: 'user-99' });
    expect(res.deleteUser).toBe(true);
  });
});

// ── updateUser ────────────────────────────────────────────────────────────────

describe('updateUser — response deserialization', () => {
  it('returns user with same id', async () => {
    const res = await client.mutations.updateUser({ id: 'user-1', input: { name: 'Updated Alice' } as UpdateUserInput });
    expect(res.updateUser.id).toBe('user-1');
  });

  it('name is updated', async () => {
    const res = await client.mutations.updateUser({ id: 'user-1', input: { name: 'Updated Alice' } as UpdateUserInput });
    expect(res.updateUser.name).toBe('Updated Alice');
  });

  it('status preserved when not in update input', async () => {
    const res = await client.mutations.updateUser({ id: 'user-1', input: { name: 'Alice Smith' } as UpdateUserInput });
    expect(res.updateUser.status).toBe(UserStatus.Active);
  });

  it('address preserved when not in update input', async () => {
    const res = await client.mutations.updateUser({ id: 'user-1', input: { name: 'Alice Smith' } as UpdateUserInput });
    expect(res.updateUser.address.street).toBe('123 Main St');
  });
});
